File: app/api/chat/utils/streamParser.ts
// File: app/api/chat/utils/streamParser.ts

import logger from 'app/api/chat/utils/logger';

/**
 * Parses a readable stream and returns the accumulated result.
 * Supports both standard JSON streams and Server-Sent Events (SSE) streams.
 * @param reader - The ReadableStream reader to parse.
 * @param options - Optional parameters to customize parsing behavior.
 * @returns {Promise<string>} - A promise that resolves with the final parsed response.
 */
interface ParseStreamOptions {
  isSSE?: boolean;      // Flag to indicate if the stream uses SSE format
  doneSignal?: string;  // The key or value indicating the stream is done
}

export async function parseStream(
  reader: ReadableStreamDefaultReader<Uint8Array>,
  options: ParseStreamOptions = { isSSE: false, doneSignal: 'done' }
): Promise<string> {
  const { isSSE = false, doneSignal = 'done' } = options;
  const decoder = new TextDecoder('utf-8');
  let buffer = '';
  let response = '';
  let done = false;

  while (!done) {
    const { value, done: streamDone } = await reader.read();
    if (streamDone) break;

    const chunk = decoder.decode(value, { stream: true });
    buffer += chunk;

    if (isSSE) {
      // SSE parsing logic
      const messages = buffer.split('\n\n');
      buffer = messages.pop() || ''; // Keep the last partial message

      for (const message of messages) {
        if (message.startsWith('data: ')) {
          const data = message.replace(/^data: /, '').trim();

          if (data === '[DONE]') {
            done = true;
            break;
          }

          try {
            const parsed = JSON.parse(data);
            if (parsed.response) {
              response += parsed.response;
            }
            if (parsed[doneSignal]) {
              done = true;
              break;
            }
          } catch (err) {
            logger.error(`app/api/chat/utils/streamParser.ts - Error parsing SSE message: "${data}"`, err);
          }
        }
      }
    } else {
      // Non-SSE parsing logic (e.g., chunked transfer encoding)
      // Split the buffer by JSON object boundaries: }{ or }\n{
      let boundary = /}\s*{/.exec(buffer);
      while (boundary) {
        const boundaryIndex = boundary.index;

        // Extract complete JSON object up to the boundary and parse it
        const completeJson = buffer.slice(0, boundaryIndex + 1);
        buffer = buffer.slice(boundaryIndex + 1); // Remove parsed object from buffer

        try {
          const parsed = JSON.parse(completeJson);
          response += parsed.response || ''; // Accumulate response text
          if (parsed[doneSignal] || false) {
            done = true; // Check for done condition based on the key
          }
        } catch (error) {
          logger.error(`app/api/chat/utils/streamParser.ts - Error parsing chunk at boundary: "${completeJson}"`, error);
        }

        // Search for the next boundary in the remaining buffer
        boundary = /}\s*{/.exec(buffer);
      }
    }
  }

  // Handle any remaining buffer content that wasn't split correctly
  if (buffer.trim()) {
    try {
      const trimmedBuffer = buffer.trim();
      if (isSSE && trimmedBuffer.startsWith('data: ')) {
        const data = trimmedBuffer.replace(/^data: /, '').trim();
        if (data !== '[DONE]') {
          const parsed = JSON.parse(data);
          response += parsed.response || '';
        }
      } else {
        const parsed = JSON.parse(buffer);
        response += parsed.response || '';
      }
    } catch (error) {
      logger.error(`app/api/chat/utils/streamParser.ts - Error parsing remaining buffer: "${buffer}"`, error);
    }
  }

  return response.trim();
}

File: app/api/chat/utils/stream_parser.py
# File: app/api/chat/utils/stream_parser.py

import asyncio
import json
import logging
import re
from typing import AsyncIterator, Optional

# Configure the logger
logger = logging.getLogger(__name__)
logger.setLevel(logging.ERROR)  # Adjust the logging level as needed
handler = logging.StreamHandler()
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
handler.setFormatter(formatter)
logger.addHandler(handler)


async def parse_stream(
    reader: AsyncIterator[bytes],
    done_signal: str = 'done'
) -> str:
    """
    Parses an asynchronous byte stream and returns the accumulated result.
    Handles cases where multiple JSON objects might be in the same chunk.

    Args:
        reader (AsyncIterator[bytes]): The async iterator yielding byte chunks to parse.
        done_signal (str, optional): The key in the parsed chunk that indicates the stream is done.
                                      Defaults to 'done'.

    Returns:
        str: The final accumulated response.
    """
    buffer = ""
    response = ""
    done = False

    # Regular expression to find boundaries between JSON objects
    boundary_pattern = re.compile(r'}\s*{')

    async for chunk in reader:
        if done:
            break

        # Decode the incoming bytes to a string using UTF-8
        try:
            decoded_chunk = chunk.decode('utf-8')
        except UnicodeDecodeError as e:
            logger.error(f"Error decoding chunk: {e}")
            continue

        buffer += decoded_chunk

        while True:
            # Search for the boundary between JSON objects
            match = boundary_pattern.search(buffer)
            if not match:
                break

            boundary_index = match.start() + 1  # Position after the first '}'

            # Extract the complete JSON object up to the boundary
            complete_json = buffer[:boundary_index]
            buffer = buffer[boundary_index:]  # Remove the parsed object from the buffer

            try:
                parsed = json.loads(complete_json)
                response += parsed.get('response', '')
                done = parsed.get(done_signal, False)
            except json.JSONDecodeError as e:
                logger.error(f"Error parsing JSON object: {complete_json}. Error: {e}")

        # Optionally, you can add logic to handle very large buffers or other edge cases

    # Handle any remaining buffer content that wasn't split correctly
    if buffer.strip():
        try:
            parsed = json.loads(buffer)
            response += parsed.get('response', '')
        except json.JSONDecodeError as e:
            logger.error(f"Error parsing remaining buffer: \"{buffer}\". Error: {e}")

    return response.strip()


# Example usage
# This is a mock example demonstrating how you might use the `parse_stream` function.
# In a real scenario, `reader` would be an asynchronous byte stream, such as one obtained
# from an HTTP response using libraries like `aiohttp`.

async def mock_reader() -> AsyncIterator[bytes]:
    """A mock async generator that yields byte chunks."""
    chunks = [
        b'{"response": "Hello", "done": false}{"response": " World", "done": false}',
        b'{"response": "!", "done": true}'
    ]
    for chunk in chunks:
        await asyncio.sleep(0.1)  # Simulate asynchronous data arrival
        yield chunk

async def main():
    accumulated_response = await parse_stream(mock_reader())
    print(f"Accumulated Response: '{accumulated_response}'")

if __name__ == "__main__":
    asyncio.run(main())
